HopsanGUI/Widgets/TextEditorWidget.cpp

```diff
@@ -0,0 +1,682 @@
+    connect(this, SIGNAL(updateRequest(QRect,int)), this, SLOT(updateLineNumberArea(QRect,int)));
+    connect(this, SIGNAL(cursorPositionChanged()), this, SLOT(highlightCurrentLine()));
+
+    updateLineNumberAreaWidth(0);
+    highlightCurrentLine();
+
+    setLineWrapMode(QPlainTextEdit::NoWrap);
+
+    mpCompleter = new QCompleter(this);
+    updateAutoCompleteList();
+
+    mpCompleter->setWidget(this);
+    mpCompleter->setCompletionMode(QCompleter::PopupCompletion);
+    mpCompleter->setCaseSensitivity(Qt::CaseInsensitive);
+    QObject::connect(mpCompleter, SIGNAL(activated(QString)),
+                     this, SLOT(insertCompletion(QString)));
+}
+
+//! @brief Paint event for line number area
+void TextEditor::lineNumberAreaPaintEvent(QPaintEvent *pEvent)
+{
+    QPainter painter(mpLineNumberArea);
+    painter.fillRect(pEvent->rect(), QColor(240,240,240)/*Qt::lightGray*/);
+
+    QTextBlock block = firstVisibleBlock();
+    int blockNumber = block.blockNumber();
+    int top = int(blockBoundingGeometry(block).translated(contentOffset()).top());
+    int bottom = top + int(blockBoundingRect(block).height());
+
+    while (block.isValid() && top <= pEvent->rect().bottom())
+    {
+        if (block.isVisible() && bottom >= pEvent->rect().top())
+        {
+            QString number = QString::number(blockNumber + 1);
+            painter.setPen(Qt::black);
+            painter.drawText(0, top, mpLineNumberArea->width(), fontMetrics().height(), Qt::AlignRight, number);
+        }
+
+        block = block.next();
+        top = bottom;
+        bottom = top + (int) blockBoundingRect(block).height();
+        ++blockNumber;
+    }
+}
+
+//! @brief Computes and returns the width of the line number area
+int TextEditor::lineNumberAreaWidth()
+{
+    int digits = 1;
+    int max = qMax(1, blockCount());
+    while (max >= 10)
+    {
+        max /= 10;
+        ++digits;
+    }
+
+    int space = 3 + fontMetrics().width(QLatin1Char('9')) * digits;
+
+    return space;
+}
+
+//! @brief Updates width of line number area
+void TextEditor::updateLineNumberAreaWidth(int /* newBlockCount */)
+{
+    setViewportMargins(lineNumberAreaWidth(), 0, 0, 0);
+}
+
+
+//! @brief Highlights current line
+void TextEditor::highlightCurrentLine()
+{
+    QList<QTextEdit::ExtraSelection> extraSelections;
+
+    if (!isReadOnly())
+    {
+        QTextEdit::ExtraSelection selection;
+
+        QColor lineColor = QColor(Qt::yellow).lighter(160);
+
+        selection.format.setBackground(lineColor);
+        selection.format.setProperty(QTextFormat::FullWidthSelection, true);
+        selection.cursor = textCursor();
+        selection.cursor.clearSelection();
+        extraSelections.append(selection);
+    }
+
+    setExtraSelections(extraSelections);
+}
+
+
+
+//! @brief Update function for line number area
+void TextEditor::updateLineNumberArea(const QRect &rect, int dy)
+{
+    if (dy)
+        mpLineNumberArea->scroll(0, dy);
```

Comment:
Missing braces on these lines, also maybe clarify dy!=0, also add documentation line for dy

Reply:
Done!